crates/storage/db/src/snapshot.rs

```diff
@@ -0,0 +1,106 @@
+                /// * `nippy_jar`: Snapshot object responsible for file generation.
+                #[allow(non_snake_case)]
+                pub fn [<create_snapshot$(_ $tbl)+>]<'tx,
+                    $($tbl: Table<Key=K>,)+
+                    HashTbl: Table<Key=K>,
+                    K
+                >
+                (
+                    tx: &impl DbTx<'tx>,
+                    range: RangeInclusive<K>,
+                    dict_compression_set: Option<Vec<impl Iterator<Item = Vec<u8>>>>,
+                    row_count: usize,
+                    nippy_jar: &mut NippyJar
+                ) -> RethResult<()>
+                    where K: Key + Copy
+                {
+                    let range: RangeInclusive<RawKey<K>> = RawKey::new(*range.start())..=RawKey::new(*range.end());
+
+                    handle_compression_and_indexing::<HashTbl>(
+                        tx,
+                        dict_compression_set,
+                        nippy_jar.uses_filters().then_some(range.clone()),
+                        nippy_jar
+                    )?;
+
+                    // Creates the cursors for the columns
+                    $(
+                        let mut [< $tbl _cursor>] = tx.cursor_read::<RawTable<$tbl>>()?;
+                        let [< $tbl _iter>] = [< $tbl _cursor>]
+                            .walk_range(range.clone())?
+                            .into_iter()
+                            .map(|row| row.map(|(_key, val)| val.take()).expect("exist in database"));
+
+                    )+
+
+                    // Create the snapshot from the data
+                    let col_iterators: Vec<Box<dyn Iterator<Item = Vec<u8>>>> = vec![
+                        $(Box::new([< $tbl _iter>]),)+
+                    ];
+
+                    nippy_jar.freeze(col_iterators, row_count as u64)?;
+
+                    Ok(())
+                }
+            }
+        )+
+    };
+}
+
+fn handle_compression_and_indexing<'tx, HashTbl: Table>(
+    tx: &impl DbTx<'tx>,
+    dict_compression_set: Option<Vec<impl Iterator<Item = Vec<u8>>>>,
+    hash_range: Option<impl RangeBounds<RawKey<<HashTbl as Table>::Key>>>,
+    nippy_jar: &mut NippyJar,
+) -> RethResult<()> {
+    // Create PHF and Filter if required
+    if let Some(hash_range) = hash_range {
+        let mut cursor = tx.cursor_read::<RawTable<HashTbl>>()?;
+
+        // This might get big but ph needs the values in memory.
+        let it = cursor
+            .walk_range(hash_range)?
+            .map(|row| row.map(|(_key, value)| value.take()))
+            .collect::<Result<Vec<_>, DatabaseError>>()?;
+
+        nippy_jar.prepare_index(&it)?;
+    }
+
+    // Create compression dictionaries if required
+    if let Some(data_sets) = dict_compression_set {
+        nippy_jar.prepare_compression(data_sets)?;
+    }
+
+    Ok(())
+}
+
+generate_snapshot_func!((T1), (T1, T2), (T1, T2, T3), (T1, T2, T3, T4),);
```

Comment:
nit: let's create wrapper functions for known snapshots, something like

```rust
pub fn create_headers_snapshot(
    tx: &impl DbTx<'tx>,
    range: RangeInclusive<BlockNumber>,
    dict_compression_set: Option<Vec<impl Iterator<Item = Vec<u8>>>>,
    row_count: usize,
    nippy_jar: &mut NippyJar
) -> RethResult<()> {
    create_snapshot_T1_T2::<Headers, HeaderTD, CanonicalHeaders, BlockNumber>(tx, range, dict_compression_set, row_count, nippy_jar)
}
```

Reply:
yes, eventually. But that's for another layer imo. That wrapper will do more than that